src/Evaluator.hs

```diff
@@ -50,313 +450,217 @@ evalErrorText (EvalErrorCase loc val) =
-          eval rhs
+          let mkEnv = extend nx x (ValueSyntax v)
+                    . extend nxs xs (ValueSyntax (Syntax (Stx scs loc (List vs))))
+          in step $ Down (unCore rhs) (mkEnv env) k
         _ -> next
-    match next (SyntaxPatternList xs) rhs =
-      \case
+    match next (SyntaxPatternList xs) rhs scrutinee env k =
+      case scrutinee of
         (ValueSyntax (Syntax (Stx _ _ (List vs))))
           | length vs == length xs ->
-            withManyExtendedEnv [(n, x, (ValueSyntax v))
-                                | (n,x) <- xs
-                                | v <- vs] $
-            eval rhs
+            let vals = [ (n, x, ValueSyntax v)
+                       | (n,x) <- xs
+                       | v     <- vs
+                       ]
+            in step $ Down (unCore rhs) (vals `extends` env) k
         _ -> next
-    match _next SyntaxPatternAny rhs =
-      const (eval rhs)
+    match _next SyntaxPatternAny rhs _scrutinee env k =
+      step $ Down (unCore rhs) env k
+
+doDataCase :: SrcLoc -> Value -> [(ConstructorPattern, Core)] -> VEnv -> Kont -> EState
+doDataCase loc v0 [] env kont = Er (EvalErrorCase loc v0) env kont
+doDataCase loc v0 ((pat, rhs) : ps) env kont =
+  match (doDataCase loc v0 ps env kont) (\newEnv -> step $ Down (unCore rhs) newEnv kont) [(unConstructorPattern pat, v0)]
+  where
+    match
+      :: EState {- ^ Failure continuation -}
+      -> (VEnv -> EState) {- ^ Success continuation, to be used in an extended environment -}
+      -> [(ConstructorPatternF ConstructorPattern, Value)] {- ^ Subpatterns and their scrutinees -}
+      -> EState
+    match _fk sk [] = sk env
+    match fk sk ((CtorPattern ctor subPats, tgt) : more) =
+      case tgt of
+        ValueCtor c args
+          | c == ctor ->
+            if length subPats /= length args
+              then error $ "Type checker bug: wrong number of pattern vars for constructor " ++ show c
+              else match fk sk (zip (map unConstructorPattern subPats) args ++ more)
+        _otherValue -> fk
+    match fk sk ((PatternVar n x, tgt) : more) =
+      match fk (sk . extend n x tgt) more
+
+-- -----------------------------------------------------------------------------
+-- Top level API
+
+evaluate :: Core -> Either EState Value
+evaluate =  evaluateIn mempty
+
+evaluateIn :: VEnv -> Core -> Either EState Value
+evaluateIn e = yield . until final step . start e . unCore
+
+evaluateWithExtendedEnv :: VEnv -> [(Ident, Var, Value)] -> Core -> Either EState Value
+evaluateWithExtendedEnv env exts = evaluateIn (inserter exts)
+  where
+    inserter = foldl' (\acc (n,x,v) -> Env.insert x n v acc) env
+
+-- TODO DYG: Move to separate module
+projectError :: EState -> EvalError
+projectError (Er err _env _k) = err
+projectError _                = error "debugger: impossible"
+
+erroneousValue :: EvalError -> Value
+erroneousValue (EvalErrorCase _loc v) = v
+erroneousValue  _                    = error "erroneusValue: likely"
```

Comment:
what does "likely" mean here? the opposite of "impossible"?

Reply:
returning to this, I cannot for the life of me understand what `likely` is doing there and `erroneous` is spelled wrong and this function is missing a case for `EvalErrorIdent`